File: src/browseDB_dumpTable.pl
#!/usr/bin/env perl

use strict;	# comment this line to speed up things
use CGI qw(-debug :standard);

$|=1;
#open(STDERR, ">&STDOUT");	

use common_DB;
if(!common_DB::checkViewAccess()){exit;}
my $dbh=common_DB::connectToDB();
if(!common_DB::checkWriteExecuteAccess()){
  print "Content-type: text/html\n\n";
  print "ERROR! Not allowed to execute scripts! ".$ENV{'REMOTE_ADDR'}; 
  exit;
 }

my $tableName		=common_DB::trim(param('tableName'));
my $crtDatabase		=common_DB::trim(param('database'));
my $remoteAddr		=$ENV{'REMOTE_ADDR'};

if ($tableName eq ""){
  print "Content-type: text/html\n\n";
  print "Table name not defined!";
  exit;
}
        
my $dbHostname=common_DB::getConfigValue('//db/host');
my $dbUsername=common_DB::getConfigValue('//db/user');
my $dbPassword=common_DB::getConfigValue('//db/user/@passwd');
my $dbDatabase=common_DB::getConfigValue('//db/database');
if($crtDatabase eq '') { $crtDatabase=$dbDatabase;}


my $pid = fork();
if (not defined $pid) {
  print "Content-type: text/html\n\n";
  print "Please try again later!\n";
  exit;
}

if ($pid == 0) { # this is the child
  $|=1;
  use POSIX ":sys_wait_h";
  sleep(1); # in case child is faster
  perform();
  exit;
}

  print "Content-type: text/mysqldump\n";
  print("Content-Disposition: attachment; filename=tableDump.sql\n");
  print("Pragma: no-cache\n");
  print("Expires: 0\n\n");

  my $kid=-1; 
  my $i=0;
  while (($kid <=0) ){
	  $kid=waitpid($pid,WNOHANG);
	  print " ";
	  sleep(10);
	  $i++; if($i>=5){last;}
  }
  
exit;
##################################################################


        
sub perform(){

  if($tableName eq "*"){$tableName="";}; ##### * is to dump all tables from the default database
  my $cmd= "/usr/bin/mysqldump --single-transaction -h$dbHostname -u$dbUsername $crtDatabase  $tableName -p$dbPassword";
  print `$cmd`;
}

File: master/browseDB_dumpTable.php
<?php 
/**
*
* Copyright (c) 2012-2013 Mount Sinai Hospital, Toronto, Ontario, 
* Copyright (c) 2012-2013 DTU/CSIG Linding Lab
*
* LICENSE:
*
* This is free software; you can redistribute it
* and/or modify it under the terms of the GNU General
* Public License as published by the Free Software Foundation;
* either version 3 of the License, or (at your option) any
* later version.
*
* This software is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public
* License along with the source code.  If not, see <http://www.gnu.org/licenses/>.
*
*
*/
include_once('common_connect.php'); checkViewAccess();
	header("Content-type: application/octet-stream");
	header("Content-Disposition: attachment; filename=tableDump.sql");
	header("Pragma: no-cache");
	header("Expires: 0");

	$tableName=trim($_REQUEST['tableName']);
	if ($tableName == ""){print "Table name not defined!"; exit;}
	
        $crtDatabase=trim($_REQUEST['database']);
        
        $cfg_xml = new SimpleXMLElement(file_get_contents('.config.xml'));
                $dbHostname=$cfg_xml->xpath('//db/host');
                $dbUsername=$cfg_xml->xpath('//db/user');
                $dbPassword=$cfg_xml->xpath('//db/user/@passwd');
                $dbDatabase=$cfg_xml->xpath('//db/database');
        if($crtDatabase=='') { $crtDatabase=$dbDatabase[0];}


  if($tableName=="*"){$tableName="";}; ##### * is to dump all tables from the default database
  print "    \n";
	print shell_exec("/usr/bin/mysqldump --single-transaction -h$dbHostname[0] -u$dbUsername[0] $crtDatabase  $tableName -p$dbPassword[0]");


	
?>

